apps/rush-lib/src/logic/taskRunner/AsyncTaskQueue.ts

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { Task } from './Task';
+import { TaskStatus } from './TaskStatus';
+
+/**
+ * @internal
+ *
+ * Implmentation of the async iteration protocol for a collection of Task objects.
+ * The async iterator will wait for a task to be ready for execution, or terminate if there are no more tasks.
+ *
+ * @remarks
+ * It is imperative that the caller update the status of tasks prior to invoking `next()` on the iterator again,
```

Comment:
Maybe there should be a pseudo-state you set it to before returning the task, to avoid requiring this be done by the implementer? Something like `TaskStatus.waitingForAction` or something (not sure what a good name would be, this is just the first thing that I didn't hate)? Just seems odd to place the ownership of setting the task state outside of the thing that cares about it.

Reply:
Maybe you can pass in the state to set it to in the constructor, and set the state as they're iterated and returned from the `next()` call?

Reply:
To phrase this a bit differently, calling `next()` on the iterator is the operation that scans the queue for ready tasks and assigns them to pending iterators. The task status is not the issue, the dependency count is.

Reply:
Then I would say the comment needs a bit of updating to better reflect that. I was a bit confused and assumed it meant the literal `Task.status`